applybn/feature_generation/bn_feature_generator.py

```diff
@@ -0,0 +1,158 @@
+import random
+import numpy as np
+import pandas as pd
+from pgmpy.models import BayesianNetwork
+from pgmpy.estimators import K2Score, BayesianEstimator, BicScore
+from pgmpy.inference import VariableElimination
+
+
+class BNFeatureGenerator:
+    def __init__(self, known_structure=None):
+        self.known_structure = known_structure
+        self.bn = None
+        self.variables = None
+        self.num_classes = None
+
+    def fit(self, data, target=None, black_list=None):
+        if not isinstance(data, pd.DataFrame):
+            data = pd.DataFrame(data)
+
+        self.variables = data.columns.tolist()
+
+        if target is not None:
+            if not isinstance(target, pd.Series):
+                target = pd.Series(target)
+            self.num_classes = len(target.unique())
+
+        if self.known_structure:
+            self.bn = BayesianNetwork(self.known_structure)
+            self.bn.fit(data, estimator=BayesianEstimator)
+        else:
+            constructor = self._BayesianNetworkConstructor(data, black_list)
+            self.bn = constructor.construct_network()
+
+    def generate_features(self, data):
+        if not isinstance(data, pd.DataFrame):
+            data = pd.DataFrame(data)
+
+        inference = VariableElimination(self.bn)
+        new_features = []
+
+        for _, instance in data.iterrows():
+            instance_features = []
+            for feature in self.variables:
+                evidence = {f: instance[f] for f in self.bn.get_parents(feature)
+                            if f in instance.index}
+                try:
+                    prob = inference.query([feature], evidence=evidence)
+                    lambda_value = prob.values[instance[feature]]
+                except:
+                    lambda_value = 1 / self.num_classes if self.num_classes else 0.5
```

Comment:
Avoid a bare except. Catch specific exceptions to better handle errors and prevent silent failures.

```suggestion
try:
    prob = inference.query([feature], evidence=evidence)
    lambda_value = prob.values[instance[feature]]
except KeyError:
    # Handle missing feature index
    lambda_value = 1 / self.num_classes if self.num_classes else 0.5
except Exception as e:
    # Fallback for any other inference error
    lambda_value = 1 / self.num_classes if self.num_classes else 0.5
    # Optionally log or raise
```